File: tests/Wording.Core.Tests/WordSelectorTests.cs
using Wording.Core;
using Wording.Core.Learning;

namespace Wording.Core.Tests;

public class WordSelectorTests
{
    static readonly DateTimeOffset Teraz = Fixtures.Teraz;

    static Word Slowko(string original, DateTimeOffset? due = null, bool reviewed = true) => new()
    {
        Original = original,
        Translation = original + "-pl",
        CreatedUtc = Teraz,
        Review = new ReviewState
        {
            DueUtc = due ?? Teraz,
            LastReviewedUtc = reviewed ? Teraz.AddDays(-1) : null,
            Repetitions = reviewed ? 1 : 0,
        },
    };

    [Fact]
    public void PustaLista_ZwracaNull()
    {
        Assert.Null(WordSelector.PickNext([], Teraz, new Random(1234)));
    }

    [Fact]
    public void JednoSlowko_ZwracaJe()
    {
        var word = Slowko("scope");

        Assert.Same(word, WordSelector.PickNext([word], Teraz, new Random(1234)));
    }

    [Fact]
    public void ZawszeZwracaSlowoZPrzekazanejListy()
    {
        var words = new[] { Slowko("a"), Slowko("b"), Slowko("c") };
        var random = new Random(1234);

        for (var i = 0; i < 500; i++)
        {
            Assert.Contains(WordSelector.PickNext(words, Teraz, random), words);
        }
    }

    [Fact]
    public void NoweSlowko_MaWyzszaWageNizSlowkoWTerminie()
    {
        var fresh = Slowko("nowe", reviewed: false);
        var due = Slowko("wterminie", Teraz);

        Assert.True(WordSelector.Weight(fresh, Teraz) > WordSelector.Weight(due, Teraz));
    }

    [Fact]
    public void Waga_RosnieWrazZOpoznieniem()
    {
        var onTime = WordSelector.Weight(Slowko("swieze", Teraz), Teraz);
        var oneDay = WordSelector.Weight(Slowko("dzien", Teraz.AddDays(-1)), Teraz);
        var oneWeek = WordSelector.Weight(Slowko("tydzien", Teraz.AddDays(-7)), Teraz);

        Assert.True(onTime < oneDay);
        Assert.True(oneDay < oneWeek);
    }

    [Fact]
    public void Waga_JestOgraniczonaZGory()
    {
        // Jedno zapomniane slowko sprzed lat nie moze zdominowac calej rotacji.
        var year = WordSelector.Weight(Slowko("stare", Teraz.AddYears(-1)), Teraz);
        var decade = WordSelector.Weight(Slowko("bardzostare", Teraz.AddYears(-10)), Teraz);

        Assert.Equal(year, decade);
    }

    [Fact]
    public void SlowkoNiewymagalne_MaMalaAleNiezerowaWage()
    {
        var weight = WordSelector.Weight(Slowko("znane", Teraz.AddDays(30)), Teraz);

        Assert.True(weight > 0, "dobrze znane slowko nie moze wypasc z rotacji calkiem");
        Assert.True(weight < WordSelector.DueWeight);
    }

    [Fact]
    public void PrzeterminowaneJestLosowaneZnaczniCzesciejNizSwiezoPowtorzone()
    {
        var overdue = Slowko("zapomniane", Teraz.AddDays(-10));
        var known = Slowko("znane", Teraz.AddDays(30));
        var words = new[] { overdue, known };
        var random = new Random(1234);

        var hits = 0;
        const int Attempts = 2000;

        for (var i = 0; i < Attempts; i++)
        {
            if (ReferenceEquals(WordSelector.PickNext(words, Teraz, random), overdue))
            {
                hits++;
            }
        }

        // Wagi to ok. 11.0 vs 0.032, wiec przeterminowane powinno dostac
        // grubo ponad 90% pokazow. Luzny prog, zeby test nie byl kruchy.
        Assert.True(hits > Attempts * 0.9, $"przeterminowane trafilo {hits}/{Attempts} razy");
    }

    [Fact]
    public void PrzyRownychWagach_RozkladJestZblizonyDoJednostajnego()
    {
        var words = new[] { Slowko("a"), Slowko("b"), Slowko("c") };
        var random = new Random(1234);
        var counts = new Dictionary<string, int> { ["a"] = 0, ["b"] = 0, ["c"] = 0 };

        for (var i = 0; i < 3000; i++)
        {
            counts[WordSelector.PickNext(words, Teraz, random)!.Original]++;
        }

        foreach (var (word, count) in counts)
        {
            Assert.True(count is > 800 and < 1200, $"{word} trafilo {count} razy, oczekiwano ok. 1000");
        }
    }
}

File: macos/Tests/WordingKitTests/WordSelectorTests.swift
import Foundation
import Testing

@testable import WordingKit

@Suite struct WordSelectorTests {
    static let now = Fixtures.now

    static func word(_ original: String, due: Date? = nil, reviewed: Bool = true) -> Word {
        Word(
            original: original,
            translation: original + "-translated",
            createdUtc: now,
            review: ReviewState(
                repetitions: reviewed ? 1 : 0,
                dueUtc: due ?? now,
                lastReviewedUtc: reviewed ? now.addingTimeInterval(-.day) : nil
            )
        )
    }

    @Test func emptyListReturnsNil() {
        var generator = SeededGenerator(seed: 1)

        #expect(WordSelector.pickNext(from: [], now: Self.now, using: &generator) == nil)
    }

    @Test func singleWordReturnsThatWord() {
        var generator = SeededGenerator(seed: 1)
        let word = Self.word("scope")

        #expect(WordSelector.pickNext(from: [word], now: Self.now, using: &generator)?.id == word.id)
    }

    @Test func newWordOutweighsAWordThatIsExactlyDue() {
        let fresh = Self.word("new", reviewed: false)
        let due = Self.word("due", due: Self.now)

        #expect(
            WordSelector.weight(for: fresh, now: Self.now)
                > WordSelector.weight(for: due, now: Self.now)
        )
    }

    @Test func weightGrowsWithLateness() {
        let onTime = WordSelector.weight(for: Self.word("on-time", due: Self.now), now: Self.now)
        let oneDay = WordSelector.weight(
            for: Self.word("one-day", due: Self.now.addingTimeInterval(-.day)), now: Self.now)
        let oneWeek = WordSelector.weight(
            for: Self.word("one-week", due: Self.now.addingTimeInterval(-7 * .day)), now: Self.now)

        #expect(onTime < oneDay)
        #expect(oneDay < oneWeek)
    }

    @Test func weightIsCapped() {
        let year = WordSelector.weight(
            for: Self.word("old", due: Self.now.addingTimeInterval(-365 * .day)), now: Self.now)
        let decade = WordSelector.weight(
            for: Self.word("ancient", due: Self.now.addingTimeInterval(-3650 * .day)), now: Self.now)

        #expect(year == decade)
    }

    @Test func wordNotYetDueKeepsASmallButNonZeroWeight() {
        let weight = WordSelector.weight(
            for: Self.word("known", due: Self.now.addingTimeInterval(30 * .day)), now: Self.now)

        #expect(weight > 0)
        #expect(weight < WordSelector.dueWeight)
    }

    @Test func overdueWordIsDrawnFarMoreOften() {
        let overdue = Self.word("forgotten", due: Self.now.addingTimeInterval(-10 * .day))
        let known = Self.word("known", due: Self.now.addingTimeInterval(30 * .day))
        let words = [overdue, known]

        var generator = SeededGenerator(seed: 42)
        var hits = 0
        let attempts = 2000

        for _ in 0..<attempts {
            if WordSelector.pickNext(from: words, now: Self.now, using: &generator)?.id == overdue.id {
                hits += 1
            }
        }

        #expect(hits > Int(Double(attempts) * 0.9))
    }

    @Test func withEqualWeightsTheDistributionIsRoughlyUniform() {
        let words = [Self.word("a"), Self.word("b"), Self.word("c")]
        var generator = SeededGenerator(seed: 7)
        var counts: [String: Int] = ["a": 0, "b": 0, "c": 0]

        for _ in 0..<3000 {
            let picked = WordSelector.pickNext(from: words, now: Self.now, using: &generator)!
            counts[picked.original, default: 0] += 1
        }

        for (word, count) in counts {
            #expect(count > 800 && count < 1200, "\(word) was picked \(count) times")
        }
    }
}
